skimage/feature/fisher_vector.py

```diff
@@ -0,0 +1,262 @@
+        vectors, and a K-mode GMM, the Fisher vector dimensionality will be
+        2KD + K. Thus, its dimensionality is invariant to the number of
+        descriptors/vectors.
+
+    References
+    -----------
+    .. [1] Perronnin, F. and Dance, C. Fisher kernels on Visual Vocabularies
+           for Image Categorization, IEEE Conference on Computer Vision and
+           Pattern Recognition, 2007
+    .. [2] Perronnin, F. and Sanchez, J. and Mensink T. Improving the Fisher
+           Kernel for Large-Scale Image Classification, ECCV, 2010
+
+    Examples
+    --------
+    >>> from skimage.feature import fisher_vector, learn_gmm
+    >>> sift_for_images = [np.random.random((10, 128)) for _ in range(10)]
+    >>> num_modes = 16
+    >>> # Estimate 16-mode GMM with these synthetic SIFT vectors
+    >>> gmm = learn_gmm(sift_for_images, n_modes=num_modes)
+    >>> test_image_descriptors = np.random.random((25, 128))
+    >>> # Compute the Fisher vector
+    >>> fv = fisher_vector(test_image_descriptors, gmm)
+    """
+    try:
+        from sklearn.mixture import GaussianMixture
+    except ImportError:
+        raise ImportError(
+            'scikit-learn is not installed. Please ensure it is installed in '
+            'order to use the Fisher vector functionality.'
+        )
+
+    if not isinstance(descriptors, np.ndarray):
+        raise DescriptorException(
+            'Please ensure descriptors is a NumPy array.'
+        )
+
+    if not isinstance(gmm, GaussianMixture):
+        raise FisherVectorException(
+            'Please ensure gmm is a sklearn.mixture.GaussianMixture object.'
+        )
+
+    if improved and not isinstance(alpha, float):
+        raise FisherVectorException(
+            'Please ensure that the alpha parameter is a float.'
+        )
+
+    num_descriptors = len(descriptors)
+
+    mixture_weights = gmm.weights_
+    means = gmm.means_
+    covariances = gmm.covariances_
+
+    posterior_probabilities = gmm.predict_proba(descriptors)
+
+    # Statistics necessary to compute GMM gradients wrt its parameters
+    pp_sum = posterior_probabilities.mean(axis=0, keepdims=True).T
+    pp_x = posterior_probabilities.T.dot(descriptors) / num_descriptors
+    pp_x_2 = posterior_probabilities.T.dot(
+        np.power(descriptors, 2)
+    ) / num_descriptors
+
+    # Compute GMM gradients wrt its parameters
+    d_pi = pp_sum.squeeze() - mixture_weights
+
+    d_mu = (pp_x - pp_sum * means).ravel()
+
+    d_sigma_t1 = pp_sum * np.power(means, 2)
+    d_sigma_t2 = pp_sum * covariances
+    d_sigma_t3 = 2 * pp_x * means
+    d_sigma = (-pp_x_2 - d_sigma_t1 + d_sigma_t2 + d_sigma_t3).ravel()
+
+    # Concatenate GMM gradients to form Fisher vector representation
+    fisher_vector = np.hstack((d_pi, d_mu, d_sigma))
+
+    if improved:
+        fisher_vector = fisher_vector / np.linalg.norm(fisher_vector)
+        fisher_vector = \
+            np.sign(fisher_vector) * np.power(np.abs(fisher_vector), alpha)
```

Comment:
There is summary of the algorithm (Algorithm 1) in this overview article by the same group: [Image Classification with the Fisher Vector: Theory and Practice](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&ved=2ahUKEwj3weStuprxAhXFG80KHcSCBmUQFnoECAMQAA&url=https%3A%2F%2Fhal.inria.fr%2Fhal-00830491v2%2Fdocument&usg=AOvVaw2jdg_Uf8o3lkJvCYBICysM)

It looks to me like there may be some subtle differences here vs. what is described:
1.) the Fisher vector computations look similar, but I do not see the divisors shown in Eqs. 31-33?
2.) it looks like the normalizations in section 3 of the algorithm are applied in the opposite order?

I have basically no prior expertise with this type of feature detection algorithm, so it may just be my misunderstanding. I see [here that the specific normalization may not matter](https://gist.github.com/danoneata/9927923/#gistcomment-2581199)?. I see from the comments there that there is actually a different reference (Krapac2011) that may correspond more closely to the specific implementation here.

The code for the implementation and the example you provided look good to me. I just don't have a great way of fully judging the correctness here and was curious whether you have tested if the implementation vs. some other reference implementation of the algorithm to see if performance is similar?

Reply:
Yes, you are correct that the implementation as it stands does not include those normalization factors from Sanchez et al. The reason being is that prior to the "improved" normalizations (L2 and power normalization), a final transformation is required for the Fisher vectors using the (inverse of the) FIM (Fisher information matrix) as the transformation matrix to each Fisher vector.

Since computing the inverse of the FIM is often intractable in practise, Sanchez et al. propose 3 approximations for this step. The first is approximate the FIM as the identity matrix, which equates to no normalization (as in the current implementation). The other two are an analytical, closed-form approximation of the FIM, and an empirical estimate of the FIM using all Fisher vectors in the training set. Since the latter empirical approximation of the FIM requires all training Fisher vectors, it is somewhat more cumbersome and less user-friendly to compute for such an implementation in scikit-image.

I have therefore added the analytical approximation normalizing scaling factors to the code (as in Sanchez et al.).

A while back I used both this implementation of the Fisher vector, as well as vlfeat's implementation for some research I was doing. I recall them performing similarly, but I eventually opted for this sklearn implementation as it suitable my Python ecosystem better at the time. I don't have those exact metrics anymore, though.

Reply:
I have also swapped the order of the L2 and power normalization.